substrate/frame/system/src/extensions/check_weight.rs

```diff
@@ -201,63 +208,56 @@ where
 	Ok(all_weight)
 }
 
-impl<T: Config + Send + Sync> SignedExtension for CheckWeight<T>
+impl<T: Config + Send + Sync> TransactionExtensionBase for CheckWeight<T> {
+	const IDENTIFIER: &'static str = "CheckWeight";
+	type Implicit = ();
+
+	fn weight() -> Weight {
+		<T::ExtensionsWeightInfo as super::WeightInfo>::check_weight()
+	}
+}
+impl<T: Config + Send + Sync, Context> TransactionExtension<T::RuntimeCall, Context>
+	for CheckWeight<T>
 where
 	T::RuntimeCall: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo>,
 {
-	type AccountId = T::AccountId;
-	type Call = T::RuntimeCall;
-	type AdditionalSigned = ();
 	type Pre = ();
-	const IDENTIFIER: &'static str = "CheckWeight";
-
-	fn additional_signed(&self) -> core::result::Result<(), TransactionValidityError> {
-		Ok(())
-	}
-
-	fn pre_dispatch(
-		self,
-		_who: &Self::AccountId,
-		_call: &Self::Call,
-		info: &DispatchInfoOf<Self::Call>,
-		len: usize,
-	) -> Result<(), TransactionValidityError> {
-		Self::do_pre_dispatch(info, len)
-	}
+	type Val = u32; /* next block length */
 
 	fn validate(
 		&self,
-		_who: &Self::AccountId,
-		_call: &Self::Call,
-		info: &DispatchInfoOf<Self::Call>,
-		len: usize,
-	) -> TransactionValidity {
-		Self::do_validate(info, len)
-	}
-
-	fn pre_dispatch_unsigned(
-		_call: &Self::Call,
-		info: &DispatchInfoOf<Self::Call>,
+		origin: T::RuntimeOrigin,
+		_call: &T::RuntimeCall,
+		info: &DispatchInfoOf<T::RuntimeCall>,
 		len: usize,
-	) -> Result<(), TransactionValidityError> {
-		Self::do_pre_dispatch(info, len)
+		_context: &mut Context,
+		_self_implicit: Self::Implicit,
+		_inherited_implication: &impl Encode,
+	) -> ValidateResult<Self::Val, T::RuntimeCall> {
+		let (validity, next_len) = Self::do_validate(info, len)?;
+		Ok((validity, next_len, origin))
 	}
 
-	fn validate_unsigned(
-		_call: &Self::Call,
-		info: &DispatchInfoOf<Self::Call>,
+	fn prepare(
+		self,
+		val: Self::Val,
+		_origin: &T::RuntimeOrigin,
+		_call: &T::RuntimeCall,
+		info: &DispatchInfoOf<T::RuntimeCall>,
 		len: usize,
-	) -> TransactionValidity {
-		Self::do_validate(info, len)
+		_context: &Context,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		Self::do_prepare(info, len, val)
 	}
 
-	fn post_dispatch(
-		_pre: Option<Self::Pre>,
-		info: &DispatchInfoOf<Self::Call>,
-		post_info: &PostDispatchInfoOf<Self::Call>,
+	fn post_dispatch_details(
+		_pre: Self::Pre,
+		info: &DispatchInfoOf<T::RuntimeCall>,
+		post_info: &PostDispatchInfoOf<T::RuntimeCall>,
 		_len: usize,
 		_result: &DispatchResult,
-	) -> Result<(), TransactionValidityError> {
+		_context: &Context,
+	) -> Result<Option<Weight>, TransactionValidityError> {
 		let unspent = post_info.calc_unspent(info);
 		if unspent.any_gt(Weight::zero()) {
 			crate::BlockWeight::<T>::mutate(|current_weight| {
```

Comment:
This extension is hard. Because `post_info` here doesn't contain any of the weight of extensions later in the pipeline. So it is underestimating the actual weight of the extrinsic and reduce too much.

One solution would be too split this extension into 2: `CheckWeight` and `CheckWeightFinalStep` (or maybe call it set_accurate_weight, or a better name).
`CheckWeight` would take into account the weight of `CheckWeightFinalStep` in its weight (thus the weight of `CheckWeightFinalStep` is zero.
And `CheckWeightFinalStep` would be put after all non-zero weight transaction extension.

Another simpler solution is to have `unspent = post_info.calc_unspent(info) - Self::weight();`
And the `CheckWeight` should be put after all the extensions with non-zero (or negligible) weight.

Reply:
Maybe this along with the other issues regarding the post dispatch weight of extensions suggest that we should be taking a different approach, where, after the call is dispatched and a `PostDispatchInfo` of the call is returned, we add to it the extension's worst case weight and instead of accruing along the pipeline, we calculate the unspent and subtract. We simply have too many extensions competing to be "the last one in the pipeline".

Reply:
but it looks doable:
* `ChargeTransaction` should be after any extension with significant refund, but no harm if missplaced.

* `CheckWeightFinalStep` is not very dependant on `CheckWeight` it can be renamed `ActuateBlockWeight`. This one should be before any extension with some post dispatch details. But no harm if missplaced, just a bit inaccurate, as the block is filled with a little more than what is actually used. But it should be after `ChargeTransaction` so that at least it is payed for.

* `StorageWeightReclaim` it is actually readjusting the work done by `ActuateBlockWeight` and use a runtime extension. This one need to strictly be last. Honestly it could also be merge with `ActuateBlockWeight` so people use either `ActuateBlockWeight` or `ActuateBlockWeightWithProofSizeExt` as final transaction extension.

Reply:
thinking again: why does `prepare` simply return `Option<Weight>` the "post dispatch details" of `validate + prepare` is there so many instance where the "post dispatch details" is only known after the call is dispatched?

It seems we can just have `prepare` returns the unspent weight and so `post_dispatch_hook` can do its post processing with full information

Reply:
> why does prepare simply return Option<Weight> the "post dispatch details" of validate + prepare is there so many instance where the "post dispatch details" is only known after the call is dispatched?

Yes, there are extensions which *can* do significant processing in their `post_dispatch`. For example, the transaction payment extensions which do asset conversion on top. They work with both native and foreign assets, with the asset path being the most expensive (assumed worst case).

As this is a core substrate primitive, we also have to think of more general use cases outside of what we build in our "standard" extensions and accomodate for significant processing during `post_dispatch`.

Reply:
> but it looks doable

It does, but it also looks pretty complicated with lots of assumptions built into these extensions. Obviously there are and will always be assumptions anyway, but my feeling is that this is probably worse for people building with these extensions than the approach with subtracting the unspent from a preloaded `PostDispatchInfo`.

Reply:
Yes great.

sidenote: also now that extensions have weight, updating the block weight could be done in `note_applied_extrinsic` in frame-system.

But it can be a breaking change, better to do in another PR.